File: app/Models/Booking.php
<?php

namespace App\Models;

use App\Enums\BookingStatus;
use App\Models\Concerns\BelongsToTenant;
use Illuminate\Database\Eloquent\Concerns\HasUlids;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;

class Booking extends Model
{
    use BelongsToTenant, HasUlids;

    protected $fillable = [
        'branch_id', 'customer_id', 'booking_date', 'start_time', 'end_time', 'status',
        'subtotal', 'discount', 'tax', 'total', 'notes', 'cancellation_reason',
        'cancelled_at', 'cancelled_by', 'created_by',
        // Phase 12 pricing breakdown/snapshot — see BookingPricingService and
        // "Booking snapshot" in LOYALTY_MEMBERSHIP_COUPON_ARCHITECTURE.md.
        // `discount` above remains the combined total of the three below,
        // for backward compatibility with the existing Flutter Booking model.
        'coupon_id', 'coupon_code', 'coupon_discount',
        'customer_membership_id', 'membership_discount',
        'loyalty_points_redeemed', 'loyalty_discount', 'loyalty_points_earned',
    ];

    protected function casts(): array
    {
        return [
            'booking_date' => 'date:Y-m-d',
            'status' => BookingStatus::class,
            'subtotal' => 'decimal:2',
            'discount' => 'decimal:2',
            'tax' => 'decimal:2',
            'total' => 'decimal:2',
            'cancelled_at' => 'datetime',
            'coupon_discount' => 'decimal:2',
            'membership_discount' => 'decimal:2',
            'loyalty_points_redeemed' => 'integer',
            'loyalty_discount' => 'decimal:2',
            'loyalty_points_earned' => 'integer',
        ];
    }

    public function tenant(): BelongsTo
    {
        return $this->belongsTo(Tenant::class);
    }

    public function branch(): BelongsTo
    {
        return $this->belongsTo(Branch::class);
    }

    public function customer(): BelongsTo
    {
        return $this->belongsTo(Customer::class);
    }

    public function coupon(): BelongsTo
    {
        return $this->belongsTo(Coupon::class);
    }

    public function customerMembership(): BelongsTo
    {
        return $this->belongsTo(CustomerMembership::class);
    }

    public function items(): HasMany
    {
        return $this->hasMany(BookingItem::class);
    }

    public function statusHistories(): HasMany
    {
        return $this->hasMany(BookingStatusHistory::class);
    }

    public function cancelledBy(): BelongsTo
    {
        return $this->belongsTo(User::class, 'cancelled_by');
    }

    public function createdBy(): BelongsTo
    {
        return $this->belongsTo(User::class, 'created_by');
    }
}

File: mobile/lib/features/booking/data/models/booking.dart
import '../../../profile/data/models/customer_profile.dart';
import 'booking_item.dart';
import 'booking_status.dart';
import 'booking_status_history_entry.dart';

/// Mirrors `App\Http\Resources\BookingResource`. `items`/`customer`/
/// `statusHistory` are only present on the backend when the controller
/// eager-loaded them (list endpoints omit them for payload size; `show`
/// endpoints include them) — see MOBILE_API_INTEGRATION.md.
class Booking {
  const Booking({
    required this.id,
    required this.branchId,
    this.customer,
    required this.bookingDate,
    required this.startTime,
    required this.endTime,
    required this.status,
    required this.subtotal,
    required this.discount,
    required this.tax,
    required this.total,
    this.couponCode,
    this.couponDiscount = 0,
    this.membershipDiscount = 0,
    this.loyaltyPointsRedeemed = 0,
    this.loyaltyDiscount = 0,
    this.loyaltyPointsEarned = 0,
    this.notes,
    this.cancellationReason,
    this.cancelledAt,
    this.items = const [],
    this.statusHistory = const [],
  });

  factory Booking.fromJson(Map<String, dynamic> json) => Booking(
    id: json['id'] as String,
    branchId: json['branch_id'] as String,
    customer: json['customer'] is Map<String, dynamic>
        ? CustomerProfile.fromJson(json['customer'] as Map<String, dynamic>)
        : null,
    bookingDate: json['booking_date'] as String,
    startTime: json['start_time'] as String,
    endTime: json['end_time'] as String,
    status: BookingStatus.fromApi(json['status'] as String),
    subtotal: num.parse(json['subtotal'].toString()),
    discount: num.parse(json['discount'].toString()),
    tax: num.parse(json['tax'].toString()),
    total: num.parse(json['total'].toString()),
    couponCode: json['coupon_code'] as String?,
    couponDiscount: num.parse((json['coupon_discount'] ?? 0).toString()),
    membershipDiscount: num.parse((json['membership_discount'] ?? 0).toString()),
    loyaltyPointsRedeemed: json['loyalty_points_redeemed'] as int? ?? 0,
    loyaltyDiscount: num.parse((json['loyalty_discount'] ?? 0).toString()),
    loyaltyPointsEarned: json['loyalty_points_earned'] as int? ?? 0,
    notes: json['notes'] as String?,
    cancellationReason: json['cancellation_reason'] as String?,
    cancelledAt: json['cancelled_at'] as String?,
    items: json['items'] is List
        ? (json['items'] as List<dynamic>).map((i) => BookingItem.fromJson(i as Map<String, dynamic>)).toList()
        : const [],
    statusHistory: json['status_history'] is List
        ? (json['status_history'] as List<dynamic>)
              .map((h) => BookingStatusHistoryEntry.fromJson(h as Map<String, dynamic>))
              .toList()
        : const [],
  );

  final String id;
  final String branchId;
  final CustomerProfile? customer;
  final String bookingDate;
  final String startTime;
  final String endTime;
  final BookingStatus status;
  final num subtotal;
  final num discount;
  final num tax;
  final num total;
  final String? couponCode;
  final num couponDiscount;
  final num membershipDiscount;
  final int loyaltyPointsRedeemed;
  final num loyaltyDiscount;
  final int loyaltyPointsEarned;
  final String? notes;
  final String? cancellationReason;
  final String? cancelledAt;
  final List<BookingItem> items;
  final List<BookingStatusHistoryEntry> statusHistory;
}
